iroh-rpc/src/client.rs

```diff
@@ -0,0 +1,377 @@
+        if outcomes.iter().any(|o| match o {
+            Ok(_) => false,
+            Err(_) => true,
+        }) {
+            return Err(RPCError::TODO);
+        };
+        Ok(())
+    }
+
+    /// Listen on a particular multiaddress
+    pub async fn listen(&mut self, addr: Multiaddr) -> Result<(), RPCError> {
+        let (sender, receiver) = oneshot::channel();
+        self.command_sender
+            .send(Command::StartListening { sender, addr })
+            .await
+            .expect("Receiver to not be dropped.");
+        match receiver.await.expect("Sender to not be dropped") {
+            SenderType::Ack => Ok(()),
+            SenderType::Error(_) => Err(RPCError::TODO),
+            _ => Err(RPCError::TODO),
+        }
+    }
+
+    /// Signal the RPC event loop to stop listening
+    pub async fn shutdown(mut self) {
+        self.command_sender
+            .send(Command::ShutDown)
+            .await
+            .expect("Receiver to still be active.");
+    }
+
+    /// Add an address and peer id associated with a particular namespace
+    pub fn with_connection_to(
+        mut self,
+        namespace: String,
+        address: Multiaddr,
+        peer_id: PeerId,
+    ) -> Self {
+        self.addresses.insert(namespace, (address, peer_id));
+        self
+    }
+
+    /// Send a single request and expects a single response
+    pub async fn call<U, V>(
+        &mut self,
+        namespace: String,
+        method: String,
+        params: U,
+    ) -> Result<V, RPCError>
+    where
+        U: Serialize + Send + Sync,
+        V: DeserializeOwned + Send + Sync,
+    {
+        let v = serialize_request(params)?;
+        let peer_id = self.get_peer_id(&namespace)?;
+        let (sender, receiver) = oneshot::channel();
+        self.command_sender
+            .send(Command::SendRequest {
+                namespace,
+                method,
+                peer_id,
+                params: v,
+                sender,
+            })
+            .await
+            .expect("Receiver not to be dropped.");
+        let res = match receiver.await.expect("Sender not to be dropped.") {
+            SenderType::Res(res) => res,
+            SenderType::Error(_) => return Err(RPCError::TODO),
+            _ => return Err(RPCError::TODO),
+        };
+        deserialize_response::<V>(&res)
+    }
+
+    /// Send a request & expects a stream of bytes as a response
+    pub async fn streaming_call<U>(
+        &mut self,
+        namespace: String,
+        method: String,
+        params: U,
+    ) -> Result<InStream, RPCError>
+    where
+        U: Serialize + Send + Sync,
+    {
+        let peer_id = self.get_peer_id(&namespace)?;
+        let v = match serialize_request(params) {
+            Ok(v) => v,
+            Err(_) => return Err(RPCError::BadRequest),
+        };
+
+        let (sender, receiver) = oneshot::channel();
+        let id: u64 = {
+            let mut rng = thread_rng();
+            rng.gen()
```

Comment:
Can't you just use a local counter for this? I don't think these need to be unique across clients